File: go/mechanisms/evm/batch-settlement/storage/channel.go
package storage

import (
	"context"
	"encoding/json"
	"errors"
	"sort"
	"sync"
	"time"

	batchsettlement "github.com/x402-foundation/x402/go/v2/mechanisms/evm/batch-settlement"
)

// Channel is the durable per-channel record shared by server and facilitator stores.
type Channel struct {
	ChannelId               string                        `json:"channelId"`
	ChannelConfig           batchsettlement.ChannelConfig `json:"channelConfig"`
	ChargedCumulativeAmount string                        `json:"chargedCumulativeAmount"`
	SignedMaxClaimable      string                        `json:"signedMaxClaimable"`
	Signature               string                        `json:"signature"`
	Balance                 string                        `json:"balance"`
	TotalClaimed            string                        `json:"totalClaimed"`
	WithdrawRequestedAt     int                           `json:"withdrawRequestedAt"`
	RefundNonce             int                           `json:"refundNonce"`
	LastRequestTimestamp    int64                         `json:"lastRequestTimestamp"`
	OnchainSyncedAt         int64                         `json:"onchainSyncedAt,omitempty"`
	Network                 string                        `json:"network,omitempty"`
}

// Base returns the channel itself so generic stores can read shared fields.
func (c *Channel) Base() *Channel { return c }

// Clone returns a shallow copy. A nil receiver yields nil so missing rows stay missing.
func (c *Channel) Clone() *Channel {
	if c == nil {
		return nil
	}
	cp := *c
	return &cp
}

// ChannelRecord is the constraint used by ChannelStorage: a record that exposes
// the shared Channel fields and can clone itself as T.
type ChannelRecord[T any] interface {
	Base() *Channel
	Clone() T
}

// ChannelUpdateStatus describes the outcome of an UpdateChannel call.
type ChannelUpdateStatus string

const (
	ChannelUpdated   ChannelUpdateStatus = "updated"
	ChannelUnchanged ChannelUpdateStatus = "unchanged"
	ChannelDeleted   ChannelUpdateStatus = "deleted"
	ChannelConflict  ChannelUpdateStatus = "conflict"
)

// ChannelUpdateResult is the result of an UpdateChannel call.
type ChannelUpdateResult[T ChannelRecord[T]] struct {
	Channel T
	Status  ChannelUpdateStatus
}

// ChannelStorage persists channel records of type T. Get returns the zero T
// (nil for pointer records) when the row is missing.
type ChannelStorage[T ChannelRecord[T]] interface {
	Get(ctx context.Context, channelId string) (T, error)
	List(ctx context.Context) ([]T, error)
	UpdateChannel(ctx context.Context, channelId string, update func(current T) T) (*ChannelUpdateResult[T], error)
}

// ChannelLockStorage is a best-effort per-channel admission lock. Loss or
// unavailability degrades to optimistic mode: the durable charge CAS still
// serializes commits. Implementation/parse errors (unreadable hold records)
// fail closed at the caller.
type ChannelLockStorage interface {
	// Acquire is SET NX + TTL. Value is pendingId. Expired keys are free.
	Acquire(ctx context.Context, channelId string, pendingId string, ttlMs int64) (bool, error)
	// Release is compare-and-delete: releases only when pendingId still holds.
	Release(ctx context.Context, channelId string, pendingId string) error
	// IsHeld reports any live lock, or this pendingId when provided (empty
	// pendingId means any live lock).
	IsHeld(ctx context.Context, channelId string, pendingId string) (bool, error)
}

// RethrowLockImplementationError returns err when it is a lock-store
// implementation/parse failure so callers fail closed.
//
// json.SyntaxError and json.UnmarshalTypeError indicate a broken backend or
// unreadable File .hold JSON. Redis lock I/O (network, timeout) is not in this
// set and stays optimistic: callers treat the lock as absent and the charge
// CAS still serializes commits.
func RethrowLockImplementationError(err error) error {
	if err == nil {
		return nil
	}
	var syntax *json.SyntaxError
	var unmarshalType *json.UnmarshalTypeError
	if errors.As(err, &syntax) || errors.As(err, &unmarshalType) {
		return err
	}
	return nil
}

// IsChannelLockStorage reports whether value implements ChannelLockStorage.
func IsChannelLockStorage(value any) bool {
	_, ok := value.(ChannelLockStorage)
	return ok
}

type admissionLock struct {
	PendingId string `json:"pendingId"`
	ExpiresAt int64  `json:"expiresAt"`
}

// InMemoryChannelStorage is a volatile ChannelStorage backed by a map.
//
// The per-channel lock map is allocated lazily and retained for the lifetime
// of the store so a caller holding the old *sync.Mutex cannot race a fresh
// lockFor allocation for the same id. The admission-lock map is dropped when
// UpdateChannel deletes the row. Long-lived servers that see an unbounded set
// of distinct channelIds will grow the lock map; production deployments
// should prefer Delete-on-drain via UpdateChannel returning the zero T.
type InMemoryChannelStorage[T ChannelRecord[T]] struct {
	mu             sync.Mutex
	sessions       map[string]T
	locks          map[string]*sync.Mutex
	admissionLocks map[string]admissionLock
}

var (
	_ ChannelStorage[*Channel] = (*InMemoryChannelStorage[*Channel])(nil)
	_ ChannelLockStorage       = (*InMemoryChannelStorage[*Channel])(nil)
)

// NewInMemoryChannelStorage creates a new in-memory channel store.
func NewInMemoryChannelStorage[T ChannelRecord[T]]() *InMemoryChannelStorage[T] {
	return &InMemoryChannelStorage[T]{
		sessions:       make(map[string]T),
		locks:          make(map[string]*sync.Mutex),
		admissionLocks: make(map[string]admissionLock),
	}
}

func (s *InMemoryChannelStorage[T]) lockFor(channelId string) *sync.Mutex {
	s.mu.Lock()
	defer s.mu.Unlock()
	lock, ok := s.locks[channelId]
	if !ok {
		lock = &sync.Mutex{}
		s.locks[channelId] = lock
	}
	return lock
}

// Get returns a clone of the stored record, or the zero T when missing.
func (s *InMemoryChannelStorage[T]) Get(_ context.Context, channelId string) (T, error) {
	var zero T
	key, err := batchsettlement.NormalizeChannelId(channelId)
	if err != nil {
		return zero, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	session, ok := s.sessions[key]
	if !ok {
		return zero, nil
	}
	return session.Clone(), nil
}

// List returns clones of every stored record, sorted by channelId so scan-backed
// queries see a stable order.
func (s *InMemoryChannelStorage[T]) List(_ context.Context) ([]T, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	result := make([]T, 0, len(s.sessions))
	for _, session := range s.sessions {
		result = append(result, session.Clone())
	}
	sort.Slice(result, func(i, j int) bool { return result[i].Base().ChannelId < result[j].Base().ChannelId })
	return result, nil
}

// UpdateChannel applies update under a per-channel mutex. Returning the same
// pointer is a no-op; returning the zero T deletes the row.
func (s *InMemoryChannelStorage[T]) UpdateChannel(_ context.Context, channelId string, update func(current T) T) (*ChannelUpdateResult[T], error) {
	key, err := batchsettlement.NormalizeChannelId(channelId)
	if err != nil {
		return nil, err
	}
	lock := s.lockFor(key)
	lock.Lock()
	defer lock.Unlock()

	s.mu.Lock()
	current, exists := s.sessions[key]
	var currentCopy T
	if exists {
		currentCopy = current.Clone()
	}
	s.mu.Unlock()

	next := update(currentCopy)
	if sameRecord(next, currentCopy) {
		return &ChannelUpdateResult[T]{Channel: currentCopy, Status: ChannelUnchanged}, nil
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	if isZeroRecord(next) {
		if exists {
			delete(s.sessions, key)
			delete(s.admissionLocks, key)
			return &ChannelUpdateResult[T]{Status: ChannelDeleted}, nil
		}
		return &ChannelUpdateResult[T]{Status: ChannelUnchanged}, nil
	}
	stored := next.Clone()
	s.sessions[key] = stored
	return &ChannelUpdateResult[T]{Channel: stored.Clone(), Status: ChannelUpdated}, nil
}

// Acquire takes a per-channel admission lock. It is not re-entrant: a live
// hold, including one owned by the same pendingId, is a miss. Refreshing TTL
// for a matching pendingId races and can extend another holder's lock.
func (s *InMemoryChannelStorage[T]) Acquire(_ context.Context, channelId string, pendingId string, ttlMs int64) (bool, error) {
	key, err := batchsettlement.NormalizeChannelId(channelId)
	if err != nil {
		return false, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	now := time.Now().UnixMilli()
	if current, ok := s.admissionLocks[key]; ok && current.ExpiresAt > now {
		return false, nil
	}
	s.admissionLocks[key] = admissionLock{PendingId: pendingId, ExpiresAt: now + ttlMs}
	return true, nil
}

// Release drops the admission lock only when pendingId still holds it.
func (s *InMemoryChannelStorage[T]) Release(_ context.Context, channelId string, pendingId string) error {
	key, err := batchsettlement.NormalizeChannelId(channelId)
	if err != nil {
		return err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if current, ok := s.admissionLocks[key]; ok && current.PendingId == pendingId {
		delete(s.admissionLocks, key)
	}
	return nil
}

// IsHeld reports whether a live admission lock exists, optionally matching pendingId.
func (s *InMemoryChannelStorage[T]) IsHeld(_ context.Context, channelId string, pendingId string) (bool, error) {
	key, err := batchsettlement.NormalizeChannelId(channelId)
	if err != nil {
		return false, err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	current, ok := s.admissionLocks[key]
	if !ok || current.ExpiresAt <= time.Now().UnixMilli() {
		delete(s.admissionLocks, key)
		return false, nil
	}
	if pendingId == "" {
		return true, nil
	}
	return current.PendingId == pendingId, nil
}

func zeroRecord[T ChannelRecord[T]]() T {
	var zero T
	return zero
}

func isZeroRecord[T ChannelRecord[T]](v T) bool {
	var zero T
	return any(v) == any(zero)
}

func sameRecord[T ChannelRecord[T]](a, b T) bool {
	return any(a) == any(b)
}

File: typescript/packages/mechanisms/evm/src/batch-settlement/storage/channel.ts
import type { ChannelConfig } from "../types";
import { normalizeChannelId } from "../utils";
import type {
  ChannelQuery,
  ChannelStoreOptions,
  QueryPage,
  SettleQuery,
  SettleTarget,
} from "./query";

export type {
  ChannelQuery,
  ChannelStoreOptions,
  QueryPage,
  SettleQuery,
  SettleTarget,
} from "./query";
export {
  matchesChannelQuery,
  queryByScan,
  queryChannels,
  querySettleTargets,
  settleQueryByScan,
  sortChannels,
} from "./query";

export interface Channel {
  channelId: string;
  channelConfig: ChannelConfig;
  chargedCumulativeAmount: string;
  signedMaxClaimable: string;
  signature: string;
  balance: string;
  totalClaimed: string;
  withdrawRequestedAt: number;
  refundNonce: number;
  onchainSyncedAt?: number;
  lastRequestTimestamp: number;
}

export interface ChannelUpdateResult<T extends Channel = Channel> {
  channel: T | undefined;
  status: "updated" | "unchanged" | "deleted" | "conflict";
}

export interface ChannelStorage<T extends Channel = Channel> {
  get(channelId: string): Promise<T | undefined>;
  list(): Promise<T[]>;
  /**
   * Atomically inspects and mutates a channel record.
   *
   * Implementations must guarantee that no concurrent mutation can interleave between
   * reading `current` and writing the callback result for all application instances that
   * share the backend. The in-memory backend only provides this guarantee inside one JS
   * runtime; production multi-instance deployments need storage with backend-level atomic
   * conditional mutation, such as Redis/Valkey Lua scripts, SQL transactions, or Durable Objects.
   *
   * @param channelId - The channel identifier.
   * @param update - Mutation callback. Return `undefined` to delete, or `current` to leave unchanged.
   * @returns The final stored channel and whether storage updated, stayed unchanged, deleted, or lost a CAS race.
   *
   * Implementations may retry compare-and-write internally. When a mutation still cannot be applied,
   * return `{ status: "conflict" }` rather than throwing. The update callback must be synchronous
   * and deterministic so retries can safely re-run on a fresher `current`.
   *
   * The compare predicate is adapter-defined: a document revision, an `expectedCharged` field check,
   * or a full-document compare are all valid; this interface does not mandate one shape.
   */
  updateChannel(
    channelId: string,
    update: (current: T | undefined) => T | undefined,
  ): Promise<ChannelUpdateResult<T>>;
  /**
   * Optional indexed worker query. Omit it and {@link queryChannels} falls back
   * to {@link queryByScan}, which dumps `list()` then filters.
   *
   * Native adapters should honour `limit` / `cursor` as read bounds. The scan
   * shim only pages after loading every row.
   *
   * @param filter - Closed named query the managers actually run.
   * @param opts - Optional per-call store options.
   * @returns One page of matching channels.
   */
  query?(filter: ChannelQuery, opts?: ChannelStoreOptions): Promise<QueryPage<T>>;
  /**
   * Optional indexed settle-target query. Omit it and {@link querySettleTargets}
   * falls back to {@link settleQueryByScan}.
   *
   * @param filter - Settle query.
   * @param opts - Optional per-call store options.
   * @returns One page of distinct claimed `(network, receiver, token)` tuples.
   */
  settleQuery?(filter: SettleQuery, opts?: ChannelStoreOptions): Promise<QueryPage<SettleTarget>>;
}

/**
 * Best-effort per-channel admission lock. Loss or unavailability degrades to
 * optimistic mode: the durable charge CAS still serializes commits.
 */
export interface ChannelLockStorage {
  /** SET NX + TTL. Value is `pendingId`. Expired keys are free. */
  acquire(channelId: string, pendingId: string, ttlMs: number): Promise<boolean>;
  /** Compare-and-delete: releases only when `pendingId` still holds. */
  release(channelId: string, pendingId: string): Promise<void>;
  /** Any live lock, or this `pendingId` when provided. */
  isHeld(channelId: string, pendingId?: string): Promise<boolean>;
}

/**
 * Rethrows lock-store implementation/parse failures so callers fail closed.
 *
 * {@link TypeError}, {@link SyntaxError}, and {@link RangeError} indicate a broken
 * backend or unreadable hold record (including corrupt File `.hold` JSON). Redis
 * lock I/O (network, timeout) is not in this set and stays optimistic: callers
 * treat the lock as absent and the charge CAS still serializes commits.
 *
 * @param err - Error from acquire, release, or isHeld.
 */
export function rethrowLockImplementationError(err: unknown): void {
  if (err instanceof TypeError || err instanceof SyntaxError || err instanceof RangeError) {
    throw err;
  }
}

/**
 * Returns whether `value` implements {@link ChannelLockStorage}.
 *
 * @param value - Storage object to inspect.
 * @returns Whether acquire/release/isHeld are present.
 */
export function isChannelLockStorage(value: object): value is ChannelLockStorage {
  return (
    "acquire" in value &&
    typeof value.acquire === "function" &&
    "release" in value &&
    typeof value.release === "function" &&
    "isHeld" in value &&
    typeof value.isHeld === "function"
  );
}

/**
 * In-memory {@link ChannelStorage} backed by a Map keyed by `channelId`.
 */
export class InMemoryChannelStorage<T extends Channel = Channel>
  implements ChannelStorage<T>, ChannelLockStorage
{
  private readonly channels = new Map<string, T>();
  private readonly channelLocks = new Map<string, Promise<void>>();
  private readonly admissionLocks = new Map<string, { pendingId: string; expiresAt: number }>();

  /**
   * Returns the channel record for a channel, if present.
   *
   * @param channelId - The channel identifier.
   * @returns The channel record or undefined when not found.
   */
  async get(channelId: string): Promise<T | undefined> {
    return this.channels.get(normalizeChannelId(channelId));
  }

  /**
   * Lists all stored channel records.
   *
   * @returns All channel records in storage.
   */
  async list(): Promise<T[]> {
    return [...this.channels.values()];
  }

  /**
   * Atomically inspects and mutates a channel record while holding a per-channel lock.
   *
   * @param channelId - The channel identifier.
   * @param update - Mutation callback. Return `undefined` to delete, or `current` to leave unchanged.
   * @returns The final stored channel and whether storage updated, stayed unchanged, or deleted.
   */
  async updateChannel(
    channelId: string,
    update: (current: T | undefined) => T | undefined,
  ): Promise<ChannelUpdateResult<T>> {
    const key = normalizeChannelId(channelId);
    return this.withChannelLock(key, async () => {
      const current = this.channels.get(key);
      const next = update(current);

      if (next === current) {
        return { channel: current, status: "unchanged" };
      }

      if (!next) {
        this.channels.delete(key);
        this.admissionLocks.delete(key);
        return { channel: undefined, status: current ? "deleted" : "unchanged" };
      }

      this.channels.set(key, next);
      return { channel: next, status: "updated" };
    });
  }

  /**
   * Acquires a per-channel admission lock if none is live.
   *
   * @param channelId - The channel identifier.
   * @param pendingId - Request-scoped lock owner.
   * @param ttlMs - Lock time-to-live in milliseconds.
   * @returns Whether this request now holds the lock.
   */
  async acquire(channelId: string, pendingId: string, ttlMs: number): Promise<boolean> {
    const key = normalizeChannelId(channelId);
    const current = this.admissionLocks.get(key);
    const now = Date.now();
    if (current && current.expiresAt > now) {
      return false;
    }
    this.admissionLocks.set(key, { pendingId, expiresAt: now + ttlMs });
    return true;
  }

  /**
   * Releases the admission lock only when `pendingId` still holds it.
   *
   * @param channelId - The channel identifier.
   * @param pendingId - Request-scoped lock owner.
   */
  async release(channelId: string, pendingId: string): Promise<void> {
    const key = normalizeChannelId(channelId);
    if (this.admissionLocks.get(key)?.pendingId === pendingId) {
      this.admissionLocks.delete(key);
    }
  }

  /**
   * Returns whether a live admission lock exists, optionally matching `pendingId`.
   *
   * @param channelId - The channel identifier.
   * @param pendingId - When set, require this request to hold the lock.
   * @returns Whether a live lock (or this request's lock) is present.
   */
  async isHeld(channelId: string, pendingId?: string): Promise<boolean> {
    const key = normalizeChannelId(channelId);
    const current = this.admissionLocks.get(key);
    if (!current || current.expiresAt <= Date.now()) {
      this.admissionLocks.delete(key);
      return false;
    }
    return pendingId === undefined || current.pendingId === pendingId;
  }

  /**
   * Runs `fn` after any prior locked work for the same channel key has finished.
   *
   * @param key - Lowercased channel id used as the lock key.
   * @param fn - Async work to run while holding the logical per-channel lock.
   * @returns The resolved result of `fn`.
   */
  private async withChannelLock<R>(key: string, fn: () => Promise<R>): Promise<R> {
    const previous = this.channelLocks.get(key) ?? Promise.resolve();
    let release!: () => void;
    const current = new Promise<void>(resolve => {
      release = resolve;
    });
    const next = previous.catch(() => {}).then(() => current);
    this.channelLocks.set(key, next);

    await previous.catch(() => {});
    try {
      return await fn();
    } finally {
      release();
      if (this.channelLocks.get(key) === next) {
        this.channelLocks.delete(key);
      }
    }
  }
}
